pkg/client/group_mgmt_client.go

```diff
@@ -61,46 +61,63 @@ type Argument struct {
 	JobId string `json:"job_id,omitempty"`
 }
 
-// NewClient instantiates a new client to communicate with the Nimble group
-func NewClient(ipAddress, username, password, apiVersion string, waitOnJobs bool) (*GroupMgmtClient, error) {
-	// Create GroupMgmt Client
-	client := resty.New()
-	client.SetTLSClientConfig(&tls.Config{
+func newGroupMgmtClient(ipAddress, username, password, apiVersion string, waitOnJobs bool) *GroupMgmtClient {
+	// Get new resty()
+	restyClient := resty.New()
+	restyClient.SetTLSClientConfig(&tls.Config{
 		InsecureSkipVerify: true,
 	})
+
+	// Create GroupMgmt Client
 	groupMgmtClient := &GroupMgmtClient{
 		URL:       fmt.Sprintf("https://%s:5392/%s", ipAddress, apiVersion),
-		Client:    client,
+		Client:    restyClient,
 		WaitOnJob: waitOnJobs,
 		Username:  username,
 		Password:  password,
 	}
+	return groupMgmtClient
+}
 
+// NewClient instantiates a new client to communicate with the Nimble group
+func NewClient(ipAddress, username, password, apiVersion string, waitOnJobs bool) (*GroupMgmtClient, error) {
+	// Get resty client
+	groupMgmtClient := newGroupMgmtClient(ipAddress, username, password, apiVersion, waitOnJobs)
 	// Get session token
 	sessionToken, err := groupMgmtClient.login(username, password)
 	if err != nil {
 		return nil, err
 	}
 	groupMgmtClient.SessionToken = sessionToken
 
+	// Login retry counter
+	loginRetry := 0
 	// Add retryCondition
 	groupMgmtClient.Client.
-		SetRetryCount(maxOpsRetries).
 		AddRetryCondition(
 			func(resp *resty.Response, err error) bool {
+
 				if err == nil && resp.StatusCode() == 401 {
 					// login again and refreh the session token
-					sessionToken, err = groupMgmtClient.login(username, password)
-					if err != nil {
+					loginRetry++
+					if loginRetry == maxLoginRetries {
```

Comment:
This seems like it will limit the number of times a long-running client can recover from a session expiration, which would be problematic.

Reply:
We don't want to have an indefinite login retry attempt otherwise we would get into a hung state. The default value for maxLoginRetries is eq 5 . we can bump it up if required. However, this is a decent retry attempt.

Reply:
You should handle this in your app . If login attempt is exhausted and sdk returned error , you should retry in your app with updated credentials.

Reply:
The application should definitely own the retry if the credentials have changed.  Forcing the application to retry on a session expiration is defeats the purpose of the automatic token refresh logic.